chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+	handler, err := protocol.NewMultiHandler(f, []byte{1, 2, 3})
+	if err != nil {
+		panic(err)
+	}
+	c.LoopHandler(ctx, handler, n)
+	r, err := handler.Result()
+	if err != nil {
+		fmt.Println(err)
+		panic(err)
+	}
+	fmt.Println("Result :", r)
+
+	// if signing is a success we register the new value
+	merkleRoot := HashMerkleRoot(pubkey, cp)
+	c.tweakedValue = HashTweakedValue(pubkey, merkleRoot)
+	c.pubkey = pubkeyShort
+	// If new config used
+	if c.newconfig != nil {
+		c.config = c.newconfig
+		c.newconfig = nil
+	}
+
+	if idsStrings[0] == c.host.ID().String() {
+		// Only first one broadcast the transaction ?
+		// Actually all participants can broadcast the transcation. It will be the same everywhere.
+		rawtx := PrepareWitnessRawTransaction(rawTransaction, r.(taproot.Signature))
+
+		payload = "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"sendrawtransaction\", \"params\": [\"" + rawtx + "\"]}"
+		result = jsonRPC(payload)
+		if result["error"] != nil {
+			fmt.Println(result)
+			panic("failed to broadcast transaction")
+		}
+
+		fmt.Println(result)
+
+		/* Need to keep this to build next one */
+		newtxid := result["result"].(string)
+		fmt.Println("New Txid:", newtxid)
+		c.ptxid = newtxid
+	}
+}
+
+func (c *CheckpointingSub) orderParticipantsList() []string {
+	id := c.host.ID().String()
+	var ids []string
+
+	ids = append(ids, id)
+
+	for _, p := range c.topic.ListPeers() {
+		ids = append(ids, p.String())
+	}
+
+	sort.Strings(ids)
+
+	return ids
+}
+
+func (c *CheckpointingSub) formIDSlice(ids []string) party.IDSlice {
+	var _ids []party.ID
+	for _, p := range ids {
+		_ids = append(_ids, party.ID(p))
+	}
+
+	idsSlice := party.NewIDSlice(_ids)
+
+	return idsSlice
+}
+
+func (c *CheckpointingSub) prefundTaproot() error {
+	taprootAddress := PubkeyToTapprootAddress(c.pubkey)
+
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"sendtoaddress\", \"params\": [\"" + taprootAddress + "\", 50]}"
+	result := jsonRPC(payload)
+	fmt.Println(result)
+	if result == nil {
+		// Should probably not panic here
+		return errors.New("couldn't create first transaction")
+	}
+	c.ptxid = result["result"].(string)
+
+	return nil
+}
+
+func (c *CheckpointingSub) initiate(cp []byte) error {
+	pubkeyShort := GenCheckpointPublicKeyTaproot(c.config.PublicKey, cp)
+	c.pubkey = pubkeyShort
+
+	idsStrings := c.orderParticipantsList()
+
+	if idsStrings[0] == c.host.ID().String() {
+		err := c.prefundTaproot()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Save tweaked value
+	merkleRoot := HashMerkleRoot(c.config.PublicKey, cp)
+	c.tweakedValue = HashTweakedValue(c.config.PublicKey, merkleRoot)
+
+	c.init = true
+
+	return nil
+}
+
+func BuildCheckpointingSub(mctx helpers.MetricsCtx, lc fx.Lifecycle, c *CheckpointingSub) {
+	ctx := helpers.LifecycleCtx(mctx, lc)
+
+	// Ping to see if bitcoind is available
+	success := BitcoindPing()
+	if !success {
+		// Should probably not panic here
+		panic("Bitcoin node not available")
```

Comment:
We can probably emit a warning saying that checkpointing is not available because there's no connection to a Bitcoin node, but allow the node to start. In this case we should be sure that we don't start any of the listening routines or event listeners, as they may error like crazy.